### PR TITLE
Fullscreen

### PR DESCRIPTION
**Pull request by fortypercenttitanium**
#### 1. Purpose:

Fullscreens the app for host and players

#### 2. Implementation:

Utilizes browser `requestFullscreen()` API.

#### 3. Possible Issues:

Needs testing on multiple browsers and devices.

#### 4. How to Test:

1. Edit `start.js` in the backend folder. The `app.listen` method needs a second argument, a string of '0.0.0.0' so the app is available on your local network:
```js
app.listen(process.env.EXPRESS_PORT, '0.0.0.0', () =>
  // eslint-disable-next-line no-console
  console.log(
    `Express listening at http://localhost:${process.env.EXPRESS_PORT}`,
  ),
);
```
2. Ensure all devices are on your local home network
3. Lookup the internal IP address of the computer running the app and write it down.
4. In frontend.env file, replace `localhost` with the IP address you just wrote down.
5. run `npm run dev:frontend` and `npm run dev:backend` in separate terminals 
6. On the host computer, you can use localhost just fine for the app, but on devices you are testing with, open the browser and navigate to `http://<HOST_COMPUTER_IP>:3000` (inside the angle brackets, replace the the text with the ip you wrote down earlier, and remove the angle brackets).
7. Host the game on one computer, join on any devices you want to test.
8. Click JOIN or HOST, and see if the game enters fullscreen mode.

#### 5. Cleanup
- [x] I've added all TODOs needed
- [x] I've removed all unneeded comments
- [x] I've updated the snapshot tests and checked to make sure they're accurate if they changed
- [x] I've removed all unnecessary `console.log`s

### CODE REVIEW

**Comment by CatQueenCodes** [comment]
Testing - can't confirm a fix

Tried on:
-iphone7
-iPhoneSE
-iPhone11
-iPadMini
-Galaxy a51
-Huwaei P30

- JOIN button screen it is not full screen on any device

- code/name screen is not full screen on iphones but is on all other devices

-from  You've Joined The Lobby Screen forward it is not full screen on iPad or iPhone. (works on all androids)

Code_Name
game_play
join_screen
Joined_Lobby

**Comment by CatQueenCodes** [comment]
Notes from call with bender after last comment above:

image0-3
iPad X does not go away
- no way for it to work on iphone
- Nice to have for desktop would be a button to go in and out of fullscreen because once you're out of it you can't get back into full screen

**Comment by I3uckwheat** [comment]
This seems to have the noslep stuff attached, can we remove that?

**Comment by CatQueenCodes** [comment]
> This seems to have the noslep stuff attached, can we remove that? 

That's all stuff for this one. It was pics of the url bar(not full screen) in different views. 

I can delete that comment with all those pics if you want though, the last comment is most relevant

File: frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+      const state = {};
+      const dispatch = jest.fn();
+
+      // tests don't include this browser function, but we are going to manually mock/restore it anyway
+      const { requestFullscreen } = document.documentElement;
+      const { fullscreenElement } = document;
+
+      // mock the requestFullscreen API
+      document.documentElement.requestFullscreen = jest.fn();
+      document.fullscreenElement = null;
```
Comment (wise-king-sullyman) [inline frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js:183]: Are these (and the associated restores) being used here?

File: frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+      const mockOnDisabledClick = jest.fn();
+
+      // tests don't include this browser function, but we are going to manually mock/restore it anyway
+      const { requestFullscreen } = document.documentElement;
+      const { fullscreenElement } = document;
+
+      // mock the requestFullscreen API
+      document.documentElement.requestFullscreen = jest.fn();
+      document.fullscreenElement = null;
```
Comment (wise-king-sullyman) [inline frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js:216]: Are these (and the associated restores) being used here?

File: frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+        .spyOn(console, 'error')
+        .mockImplementation(() => {});
+
+      // tests don't include this browser function, but we are going to manually mock/restore it anyway
+      const { requestFullscreen } = document.documentElement;
+      const { fullscreenElement } = document;
+
+      // mock the requestFullscreen API
+      document.documentElement.requestFullscreen = jest.fn();
+      document.fullscreenElement = null;
```
Comment (wise-king-sullyman) [inline frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js:254]: Are these (and the associated restores) being used here?

File: frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
```
Comment (wise-king-sullyman) [inline frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js:284]: Are these (and the associated restores) being used here?

File: frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
```
Comment (wise-king-sullyman) [inline frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js:313]: Are these (and the associated restores) being used here?

File: frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
```
Comment (wise-king-sullyman) [inline frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js:342]: Are these (and the associated restores) being used here?

File: frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
```
Comment (wise-king-sullyman) [inline frontend/src/components/HostSettingsMenu/SettingsSubComponents/ToggleFullscreen.spec.js:371]: Are these (and the associated restores) being used here?

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
Works great, just some minor things in the tests to clean up as we discussed.

**Review by wise-king-sullyman — COMMENTED** [review]
One question, then once that's discussed/changed it should be good to go.

File: backend/start.js [hunk 8]
```diff
@@ -7,7 +7,7 @@ const app = require('./app/express.js');
 /* ======= Start socket service ======= */
 require('./app/socket.js');
 
-app.listen(process.env.EXPRESS_PORT, () =>
+app.listen(process.env.EXPRESS_PORT, '0.0.0.0', () =>
```
Comment (wise-king-sullyman) [inline backend/start.js:10]: Did you mean to leave this in? I got the feeling from your instructions this shouldn't be pushed but I could be wrong.

**Review by wise-king-sullyman — APPROVED** [review]
Looks great!

**Review by wise-king-sullyman — APPROVED** [review]

File: frontend/src/helpers/requestFullscreen.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+export default function requestFullscreen() {
+  const docEl = document.documentElement;
+  const requestFullscreenPrefix =
+    docEl.requestFullscreen ||
+    docEl.mozRequestFullScreen ||
+    docEl.webkitRequestFullScreen ||
+    docEl.msRequestFullscreen;
```
Comment (wise-king-sullyman) [inline frontend/src/helpers/requestFullscreen.js:7]: I would like to see no call happen if all of these return a falsy value

File: frontend/src/helpers/requestFullscreen.spec.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    expect(document.documentElement.webkitRequestFullScreen).toHaveBeenCalled();
+  });
+
+  it('calls msRequestFullscreen if the method exists on the documentElement object', () => {
+    global.document.documentElement.msRequestFullscreen = jest.fn();
+
+    requestFullscreen();
+
+    expect(document.documentElement.msRequestFullscreen).toHaveBeenCalled();
+  });
```
Comment (wise-king-sullyman) [inline frontend/src/helpers/requestFullscreen.spec.js:47]: And a test added here for all of the fullscreen requests being falsy.

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]

**Review by I3uckwheat — APPROVED** [review]

**Review by wise-king-sullyman — APPROVED** [review]

**Pull request merged**